src/irmin/commit.ml

```diff
@@ -25,43 +25,74 @@ module Log = (val Logs.src_log src : Logs.LOG)
 module Maker (Info : Info.S) = struct
   module Info = Info
 
-  module Make (H : Type.S) = struct
+  module Make
+      (H : Type.S)
+      (N : Key.S with type hash = H.t)
+      (C : Key.S with type hash = H.t) =
+  struct
     module Info = Info
 
-    type hash = H.t [@@deriving irmin ~compare]
+    type node_key = N.t [@@deriving irmin ~compare]
+    type commit_key = C.t [@@deriving irmin]
 
-    type t = { node : hash; parents : hash list; info : Info.t }
+    type t = { node : node_key; parents : commit_key list; info : Info.t }
```

Comment:
the distinction between `node_key` `commit_key` and `contents_key` can maybe be extracted to a separate PR?

Reply:
In some places it's possible to drop the distinction; in others (such as in `Irmin.Node.S`), it's not.

I'll do a pass at the end to try to simplify where possible :+1: